samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -119,24 +120,30 @@ public void run(StreamApplication app) {
         throw new SamzaException("No jobs to run.");
       }
       plan.getJobConfigs().forEach(jobConfig -> {
-          log.info("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
-          String processorId = getProcessorId(config);
-          StreamProcessor processor =
-              createStreamProcessor(processorId, jobConfig, app, new LocalStreamProcessorListener(processorId));
-          processor.start();
+          log.debug("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
+          LocalStreamProcessorLifeCycleListener listener = new LocalStreamProcessorLifeCycleListener();
+          StreamProcessor processor = createStreamProcessor(jobConfig, app, listener);
+          listener.setProcessor(processor);
```

Comment:
If you know the number of streamprocessors that will be spawned, can you simply maintain a countdown latch with that count, decrement it in the lifecycle callbacks and then, await on it here? 

Setting processor reference inside the listener feels unclean for me. Using startLatch and endLatch may be cleaner. Let me know if you already tried it and didn't work out for some reason.